napari/utils/settings/_manager.py

```diff
@@ -0,0 +1,156 @@
+"""Settings management.
+"""
+
+import os
+from pathlib import Path
+
+from pydantic import ValidationError
+from yaml import safe_dump, safe_load
+
+from napari.utils.settings._defaults import CORE_SETTINGS
+
+
+class SettingsManager:
+    """
+    Napari settings manager using evented SettingsModels.
+
+    This provides the presistence layer for the application settings.
+
+    Parameters
+    ----------
+    config_path : str, optional
+        Provide the base folder to store napari configuration. Default is None,
+        which will point to `"~/.napari/settings/`.
+    save_to_disk : bool, optional
+        Persist settings on disk. Default is True.
+    """
+
+    _FILENAME = "settings.yaml"
+
+    def __init__(self, config_path: str = None, save_to_disk: bool = True):
+        # TODO: Use this instead Path.home() / ".config" / ".napari" ?
+        self._config_path = (
+            Path.home() / ".napari" / "settings"
+            if config_path is None
+            else config_path
+        )
+        self._save_to_disk = save_to_disk
+        self._settings = {}
+        self._defaults = {}
+        self._models = {}
+        self._plugins = []
+
+        if not self._config_path.is_dir():
+            os.makedirs(self._config_path)
+
+        self._load()
+
+    def __getattr__(self, attr):
+        if attr in self._settings:
+            return self._settings[attr]
+
+    def __dir__(self):
+        """Add setting keys to make tab completion works."""
+        return super().__dir__() + list(self._settings)
+
+    @staticmethod
+    def _get_section_name(settings) -> str:
+        """
+        Return the normalized name of a section based on its config title.
+        """
+        section = settings.Config.title.replace(" ", "_").lower()
+        if section.endswith("_settings"):
+            section = section.replace("_settings", "")
+
+        return section
+
+    def _to_dict(self) -> dict:
+        """Convert the settings to a dictionary."""
+        data = {}
+        for section, model in self._settings.items():
+            data[section] = model.dict()
+
+        return data
+
+    def _save(self):
+        """Save configuration to disk."""
+        if self._save_to_disk:
+            path = self._config_path / self._FILENAME
+            with open(path, "w") as fh:
+                fh.write(safe_dump(self._to_dict()))
+
+    def _load(self):
+        """Read configuration from disk.
+
+        If no settings file is found, a new one is created using the defaults.
```

Comment:
It'd be nice to expand either this docstring, or the class docstring, to provide a little more overview about how this all works.  I can "glean" it from reading the code, but I'd rather have  a brief guide from the "architect" of the object.

Reply:
Will do.